pkg/persistence/postgres.go

```diff
@@ -0,0 +1,387 @@
+package persistence
+
+import (
+	"bytes"
+	"database/sql"
+	"fmt"
+	"net/url"
+	"text/template"
+	"time"
+	"unicode/utf8"
+
+	_ "github.com/lib/pq"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type postgresDatabase struct {
+	conn   *sql.DB
+	schema string
+}
+
+func makePostgresDatabase(url_ *url.URL) (Database, error) {
+	db := new(postgresDatabase)
+
+	schema := url_.Query().Get("schema")
+	if "" == schema {
+		db.schema = "magneticod"
+	} else {
+		db.schema = schema
+	}
+
+	var err error
+	db.conn, err = sql.Open("postgres", url_.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "sql.Open")
+	}
+
+	// > Open may just validate its arguments without creating a connection to the database. To
+	// > verify that the data source Name is valid, call Ping.
+	// https://golang.org/pkg/database/sql/#Open
+	if err = db.conn.Ping(); err != nil {
+		return nil, errors.Wrap(err, "sql.DB.Ping")
+	}
+
+	// https://github.com/mattn/go-sqlite3/issues/618
+	db.conn.SetConnMaxLifetime(0) // https://golang.org/pkg/database/sql/#DB.SetConnMaxLifetime
+	db.conn.SetMaxOpenConns(3)
+	db.conn.SetMaxIdleConns(3)
+
+	if err := db.setupDatabase(); err != nil {
+		return nil, errors.Wrap(err, "setupDatabase")
+	}
+
+	return db, nil
+}
+
+func (db *postgresDatabase) Engine() databaseEngine {
+	return Postgres
+}
+
+func (db *postgresDatabase) DoesTorrentExist(infoHash []byte) (bool, error) {
+	rows, err := db.conn.Query("SELECT 1 FROM "+db.schema+".torrents WHERE info_hash = $1;", infoHash)
```

Comment:
Explicitly concatenating `db.schema` is inelegant.

I believe we can use `SET search_path TO my_schema;` at the beginning of the sessions instead, to set the default schema. See https://www.postgresql.org/docs/9.3/ddl-schemas.html#DDL-SCHEMAS-PATH

What do you think?

Reply:
> I believe we can use `SET search_path TO my_schema;`

I don't think this is a good idea.

AFAIR `SET search_path` is setting `search_path` for database and not for the session. So:

- Other connections and sessions will also be affected
- It requires rights to alter database settings

When developing applications which are working with PostgreSQL I'm not using `search_path` at all and considering it a bad practice. It almost always better and more reliable to explicitly state schema before the table name. Especially when working with multi-schema databases.

Reply:
@boramalper Regardless of my concerns it's done.